### DISCUSSION TITLE
How to read link discharge in a structure (direct discharge on a river network)

### DISCUSSION THREAD

**Discussion by hawklorry** [DHI/mikeio1d#44] [Q&A] [answered]
An example would be great. I tried following code. It didn't work. 

`queries = QueryDataReach("Discharge", "Fish_Valve_June_1", 0)
df = res1d.read(queries)`

**Comment by JesperGr** [comment]
The outputs of the structure is stored in the reach at the chainage of the structure. If the structure is located at "aBranch" at chainage 450, then you should be able to grab the discharge using:

```
QueryDataReach("Discharge", "aBranch", 450)
```

**Reply by hawklorry** [reply]
Thanks @JesperGr.  As shown below, there are multiple structures at the same chainage. I'm wondering how we could specify the one that we want to use in the query. 

![image](https://user-images.githubusercontent.com/3594975/218478424-4cca4556-beb2-47bc-9bd5-c4f25628e5c0.png)

**Reply by gedaskir** [reply, marked as answer]
If it is not too big of a hassle you could try to use a development implementation of mikeio1d from **feature/result_network_auto_completion** branch.  You can install a feature branch using:

```cmd
pip install https://github.com/DHI/mikeio1d/archive/feature/result_network_auto_completion.zip
```

Here I implemented ```QueryDataStructure```, which can be used as:

```python
from mikeio1d import Res1D
from mikeio1d.query import QueryDataStructure
res1d = Res1D('MyRes1D.resd')
query_structure = QueryDataStructure('DischargeInStructure', 'MyStructureId')
df = res1d.read(query_structure)
```
For now ```QueryDataStructure``` has a drawback that it requires all structure IDs to be unique.
Additionally, that branch has auto-completion functionality for Jupyter etc., so now one could add a query using:

```python
from mikeio1d import Res1D
res1d = Res1D('MyRes1D.resd')
res1d.structures.MyStructureId.DischargeInStructure.add()
df = res1d.read()
```
This should also work for nodes, catchments, reaches. Note that this is still in development so it is not stable :-).

**Reply by gedaskir** [reply]
Now this functionality is made available as part of MIKE IO 1D v0.3.

**Reply by hawklorry** [reply]
Perfect. I will try it out. Do we also have a function to get the ids of all structures?

**Reply by gedaskir** [reply]
You should be able to get the ids of all structures using `res1d.structures`.

**Reply by hawklorry** [reply]
> You should be able to get the ids of all structures using `res1d.structures`.

It works well. Thanks for the new updates. It's a time-saver.
